snappy/security.go

```diff
@@ -20,207 +20,835 @@
+	return filepath.Join(dirs.GlobalRootDir, aaPolicyDir)
+}
+
+func scPolicyDir() string {
+	scPolicyDir := "/usr/share/seccomp"
+	return filepath.Join(dirs.GlobalRootDir, scPolicyDir)
+}
+
+func aaFrameworkPolicyDir() string {
+	aaFrameworkPolicyDir := filepath.Join(policy.SecBase, "apparmor")
+	return filepath.Join(dirs.GlobalRootDir, aaFrameworkPolicyDir)
+}
+
+func scFrameworkPolicyDir() string {
+	scFrameworkPolicyDir := filepath.Join(policy.SecBase, "seccomp")
+	return filepath.Join(dirs.GlobalRootDir, scFrameworkPolicyDir)
+}
+
+func defaultPolicyVendor() string {
+	// FIXME: slightly ugly that we have to give a prefix here
+	return fmt.Sprintf("ubuntu-%s", release.Get().Flavor)
+}
+
+func defaultPolicyVersion() string {
+	return release.Get().Series
+}
+
+var dbusPathOkRegexp = regexp.MustCompile(`^[a-zA-Z0-9]$`)
+
+// Generate a string suitable for use in a DBus object
+func dbusPath(s string) string {
+	dbusStr := ""
+
+	for _, c := range strings.SplitAfter(s, "") {
+		if dbusPathOkRegexp.MatchString(c) {
+			dbusStr += c
+		} else {
+			dbusStr += fmt.Sprintf("_%02x", c)
+		}
+	}
+
+	return dbusStr
+}
+
+// Calculate whitespace prefix based on occurrence of s in t
+func findWhitespacePrefix(t string, s string) string {
+	pat := regexp.MustCompile(`^ *` + regexp.QuoteMeta(s))
+	p := ""
+	for _, line := range strings.Split(t, "\n") {
+		if pat.MatchString(line) {
+			for i := 0; i < len(line)-len(strings.TrimLeft(line, " ")); i++ {
+				p += " "
+			}
+			break
+		}
+	}
+	return p
+}
+
+func getSecurityProfile(m *packageYaml, appName, baseDir string) (string, error) {
+	cleanedName := strings.Replace(appName, "/", "-", -1)
+	if m.Type == pkg.TypeFramework || m.Type == pkg.TypeOem {
+		return fmt.Sprintf("%s_%s_%s", m.Name, cleanedName, m.Version), nil
+	}
+
+	origin, err := originFromYamlPath(filepath.Join(baseDir, "meta", "package.yaml"))
+
+	return fmt.Sprintf("%s.%s_%s_%s", m.Name, origin, cleanedName, m.Version), err
+}
+
+func findTemplate(template string, policyType string) (string, error) {
+	if template == "" {
+		template = defaultTemplate
+	}
+
+	systemTemplate := ""
+	fwTemplate := ""
+	subdir := filepath.Join("templates", defaultPolicyVendor(), defaultPolicyVersion())
+	if policyType == "apparmor" {
+		systemTemplate = filepath.Join(aaPolicyDir(), subdir, template)
+		fwTemplate = filepath.Join(aaFrameworkPolicyDir(), "templates", template)
+	} else if policyType == "seccomp" {
+		systemTemplate = filepath.Join(scPolicyDir(), subdir, template)
+		fwTemplate = filepath.Join(scFrameworkPolicyDir(), "templates", template)
+	} else {
+		return "", errPolicyTypeNotFound
+	}
+
+	// Always prefer system policy
+	fns := []string{systemTemplate, fwTemplate}
+	for _, fn := range fns {
+		content, err := ioutil.ReadFile(fn)
+		if err == nil {
+			return string(content), nil
+		}
+	}
+
+	return "", &errPolicyNotFound{"template", policyType, template}
+}
+
+func findCaps(caps []string, template string, policyType string) (string, error) {
```

Comment:
I don't like how findCaps reads in a bunch of files, gloms them all into a string, and then everything that calls findCaps splits the returned string into lines. That seems both awkward and inefficient -- maybe I'm missing a use case that makes this make sense, but perhaps using bufio to make the returned thing be a []string directly would be better?